dss/notify/README.md

```diff
@@ -0,0 +1,93 @@
+
+In the context of the DSS Data Store, a notification is defined as the invocation of a POST request against an HTTP or
+HTTPS endpoint at the time a particular type of event occurs within the data store. The only event for which
+notifications are currently supported is that of a new bundle version beeing indexed. Registering an endpoint to be
+notified in an event of a certain type results in the creation of a persistent *subscription* object. The system will
+notify the subscribed endpoint every time such an event occurs. The subscription for a *bundle version indexed* event
+can be configured to limit notifications to new bundle versions whose metadata matches a particular ElasticSearch query.
+
+A notification is known to be synchronous if it is delivered in the context of the data store thread that handles the
+event. It is considered asynchronous if it is delivered by a separate data store thread, enabling the event handling
+thread to continue its work without having to deal with outages of the endpoint to be notified. DSS Data Store
+notifications are represented by instances of the `Notification` class. The actual queueing and retry logic for
+notifications resides in the `Notifier` class. For synchronous notifications, the `Notification.deliver()` method is
+invoked directly. For asynchronous notifications, the `Notification` instance is instead passed to the `enqueue()`
+method of a `Notifier` instance. That method wraps the `Notification` instance in an SQS message and enqueues it in an
+SQS queue.
+
+The `Notifier.run()` method starts the asynchronous notification process. It launches a number of worker threads that
+each pull a message from a queue, unwrap the `Notification` instance inside the message and invoke `deliver()` on it.
+If `deliver()` fails or times out, the message is put back onto an SQS queue where it waits for another attempt.
+
+The `Notifier.enqueue()` method is typically invoked by the `dss-index` daemon in response to an S3 event about a
+bundle that is to be indexed. The `Notifier.run()` method is typically invoked by the `dss-notify` daemon which is
+started periodically by a CloudWatch rule.
+
+There are multiple SQS queues, one per delivery attempt. The queues are numbered, starting with 0. All new
+notifications are enqueued in the queue with index 0. All notifications that failed their first delivery attempt are
+placed in queue 1 and so on. Each queue has at minimum one worker thread assigned to it. Having multiple queues
+guarantees the fairness and responsiveness of the design: newly enqueued notifications are handled with the same
+priority as older notifications that already had to be retried. The presence of many notifications that have to be
+retried will not affect the ability of the system to handle new notifications. It allows the system to keep up with
+ongoing new work, while concurrently recovering from prior outages of itself or individual notification endpoints.
```

Comment:
I'd also vote for a single queue approach. As we discussed offline the backoff rules could be implemented via ChangeMessageVisibility on a per message basis.

Reply:
I agree with the single queue, and also having a dead letter queue for expired notifications. https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html.

Reply:
The multi-queue design makes it very easy to reason about fairness. I don't have any concrete evidence that a single-queue design can't be made fair but I fear that a naive single queue design could be unfair. That's intuitively apparent if you imagine the scenario I describe in the README. OTOH, what I do *know* is that this design *is* fair.

SQS Push has the potential to change everything because it can automatically scale up worker lambda invocations, addressing some of my fairness concerns with a single queue design.